### PR TITLE
Feature: Add js-client step by step guide

### PR DESCRIPTION
**Pull request by josemarinas**

### CODE REVIEW

File: packages/js-client/GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+
+As mentioned before, the JS client is a totally optional step in the process of creating an Aragon OS plugin. You can definitely build a plugin without a JS client, but you will be missing some benefits.
+
+Aragon common packages will handle the connection to IPFS, Subgraph and blockchain providers, so you don't have to reinvent the wheel.
+
+Additionally, is giving some structure to your plugin, if you plan on doing some complex logic in your plugin, it is a good idea to have a JS client to handle it and leave the UI handling only the UI logic.
+
+## Getting started
+
+To get started with your client the first step is to create a nes JS/TS project you can use any tool you want.
```
Comment (pythonpete32) [inline packages/js-client/GUIDE.md:38]:
```suggestion
To get started with your client the first step is to create a nes JS/TS project you can use any tool you want.
```

Comment (AaronAbuUsama) [inline packages/js-client/GUIDE.md]:
```suggestion
To get started with your client the first step is to create a new JS/TS project you can use any tool you want.
```

File: packages/js-client/GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+
+Once you have your project set up yo need to install the `@aragon/sdk-client-common` package.
```
Comment (AaronAbuUsama) [inline packages/js-client/GUIDE.md]:
```suggestion
Once you have your project set up you need to install the `@aragon/sdk-client-common` package.
```

File: packages/js-client/GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+      // override the context params with the ones passed to the constructor
+      this.set(contextParams);
+    }
+  }
+
+  // Other methods
+}
+```
+
+You can also add more methods and helpers in this calss that can make your life easieer whule building your client like `getters` or `setters` for the state or a function that sets the defualt values for the state.
```
Comment (AaronAbuUsama) [inline packages/js-client/GUIDE.md]:
```suggestion
You can also add more methods and helpers in this class that can make your life easier while building your client like `getters` or `setters` for the state or a function that sets the defualt values for the state.
```

File: packages/js-client/GUIDE.md [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+    super(context);
+    this.methods = new MyCustomMethodsModule(context);
+    this.estimation = new MyCustomEstimationModule(context);
+    this.myCustomParam = context.myCustomParam;
+    this.myCustomBackendClient = context.myCustomBackendClient;
+  }
+}
+```
+
+Now when you use the client from an UI you will be able to access the modules and the variables.
```
Comment (AaronAbuUsama) [inline packages/js-client/GUIDE.md]:
```suggestion
Now when you use the client from a UI you will be able to access the modules and the variables.
```

File: packages/js-client/GUIDE.md [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+  myCustomParam: string;
+  // In the context you will be able to work with the params
+  // so the state can be a client for a service instead of the
+  // actual value
+  myCustomBackendClient: MyCustomBackendClient[];
+};
+
+export type MyPluginOverriddenState = OverriddenState & {
+  [key in keyof MyPluginContextState]: boolean;
+};
```
Comment (AaronAbuUsama) [inline packages/js-client/GUIDE.md:88]: Im a little lost here. There isnt a good explanation of what these types are or how to define them. and when you look in the template its not obvious as to what they map to. especially the back end client

File: packages/js-client/GUIDE.md [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+```
+
+### `src/context.ts`
+
+The context is the responsible for storing the state of the of the client and setting the default values for the params. The context is also responsible for doing any kind of operation such as modifying some parameter, or using a parameter for creating a client for a service.
```
Comment (AaronAbuUsama) [inline packages/js-client/GUIDE.md]: what parameters are we talking about here (internal parameters), also what does it mean to use a parameter for creating a client for a service? is this another client other than the one this context is a part of?

File: packages/js-client/GUIDE.md [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+
+```ts
+// src/context.ts
+export class MyPluginClientContext extends Context {
+  protected state: MyPluginContextState;
+  protected overridden: MyPluginOverriddenState;
+
+  constructor(params?: Partial<MyPluginClientParams>) {
+    super();
+    this.set(contextParams);
```
Comment (AaronAbuUsama) [inline packages/js-client/GUIDE.md]:
```suggestion
    this.set(params);
```

File: packages/js-client/GUIDE.md [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+  }
+
+  public set(params: Partial<MyPluginClientParams>) {
+    // First we need to set the values in the parent class
+    super.set(params);
+    // Then we can set the values in our context
+    if (params.myCustomParam) {
```
Comment (AaronAbuUsama) [inline packages/js-client/GUIDE.md:115]: so in this set function we switch between all the things we added in `MyPluginState`

File: packages/js-client/GUIDE.md (hunk 3 above)
Comment (AaronAbuUsama) [inline packages/js-client/GUIDE.md]:
```suggestion
You can also add more methods and helpers in this class that can make your life easier while building your client like `getters` or `setters` for the state or a function that sets the default values for the state.
```

File: packages/js-client/GUIDE.md [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+  protected state: MyPluginContextState;
+  protected overridden: MyPluginOverriddenState;
+
+  constructor(params?: Partial<MyPluginClientParams>, context?: Context) {
+    super();
+    if (context) {
+      // copy the context properties to this
+      Object.assign(this, context);
+    }
+    if (contextParams) {
```
Comment (AaronAbuUsama) [inline packages/js-client/GUIDE.md]:
```suggestion
    if (params) {
```

File: packages/js-client/GUIDE.md [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+      // override the context params with the ones passed to the constructor
+      this.set(contextParams);
```
Comment (AaronAbuUsama) [inline packages/js-client/GUIDE.md]:
```suggestion
      this.set(params);
```

File: packages/js-client/GUIDE.md (hunk 8 above)
Comment (josemarinas) [reply packages/js-client/GUIDE.md:115]: Yep the idea is that you check that the param exists and then you set the overriden to true

File: packages/js-client/GUIDE.md (hunk 5 above)
Comment (josemarinas) [reply packages/js-client/GUIDE.md:88]: I added some more explaining lines, please re review it

File: packages/js-client/GUIDE.md (hunk 6 above)
Comment (josemarinas) [reply packages/js-client/GUIDE.md]: I updated this paragraph

File: packages/js-client/GUIDE.md (hunk 8 above)
Comment (josemarinas) [reply packages/js-client/GUIDE.md:115]: I added some comments here so it is clearer

File: packages/js-client/GUIDE.md (hunk 5 above)
Comment (AaronAbuUsama) [reply packages/js-client/GUIDE.md:88]: Awesome!
I have not followed the the guide and built a client yet but reading through it makes sense and im no longer lost  =]

**Review by AaronAbuUsama — APPROVED** [review]

File: packages/js-client/EXAMPLE.md [hunk 11]
```diff
@@ -0,0 +1,665 @@
+# Example of plugin
```
Comment (heueristik) [inline packages/js-client/EXAMPLE.md]:
```suggestion
# Plugin Example
```

File: packages/js-client/EXAMPLE.md [hunk 12]
```diff
@@ -0,0 +1,665 @@
...
+
+This file aims to provide an example of how to build a `js-client` following the guide in the [GUIDE.md](./GUIDE.md).
+
+For this example we are going to build a `js-client` for the template plugin in this repository.
+
+## Initialize the project
```
Comment (heueristik) [inline packages/js-client/EXAMPLE.md]: https://capitalizemytitle.com/ is a helpful tool.

File: packages/js-client/EXAMPLE.md [hunk 13]
```diff
@@ -0,0 +1,665 @@
...
+
+The fist step would be initializing a `npm` project. You can use the following commands to do so ot any other tool you prefer.
```
Comment (heueristik) [inline packages/js-client/EXAMPLE.md]:
```suggestion
The fist step is to initialize a `npm` project. You can use the following commands:
```

File: packages/js-client/EXAMPLE.md [hunk 14]
```diff
@@ -0,0 +1,665 @@
...
+
+Yarn:
+
+```bash
+mkdir js-client-example
+cd js-client-example
+yarn init -y
+```
+
+Now we need to install the common tooling for `js-client` development.
```
Comment (heueristik) [inline packages/js-client/EXAMPLE.md]:
```suggestion
Now, we need to install the common tooling for `js-client` development.
```

File: packages/js-client/EXAMPLE.md [hunk 15]
```diff
@@ -0,0 +1,665 @@
...
+Yarn:
+
+```bash
+yarn add `@aragon/sdk-client-common`
+
+### Optional
+yarn add `@aragon/sdk-common`
+```
+
+Now create a `src` folder and a `index.ts` file inside it.
```
Comment (heueristik) [inline packages/js-client/EXAMPLE.md]:
```suggestion
Now, create a `src` folder and a `index.ts` file inside it.
```

File: packages/js-client/EXAMPLE.md [hunk 16]
```diff
@@ -0,0 +1,665 @@
...
+
+```bash
+mkdir src
+touch src/index.ts
+```
+
+The index file will be the entry point of our `js-client` and will export everything in the non internal folders.
```
Comment (heueristik) [inline packages/js-client/EXAMPLE.md]:
```suggestion
The index file will be the entry point of our `js-client` and will export everything in the non-internal folders.
```

File: packages/js-client/GUIDE.md [hunk 17]
```diff
@@ -0,0 +1,448 @@
...
+
+- Web3 providers
+- Subgraph
+- IPFS
+
+This package handles for you the connection (fail prevention, error handling, connection retry...) and basic functionality of these services, so you can focus on building your plugin.
+
+### Why a JS client?
+
+As mentioned before, the JS client is a totally optional step in the process of creating an Aragon OS plugin. You can definitely build a plugin without a JS client, but you will be missing some benefits.
```
Comment (heueristik) [inline packages/js-client/GUIDE.md]:
```suggestion
As mentioned before, the JS client is an optional step in the process of creating an Aragon OSx plugin. You can definitely build a plugin without a JS client, but you will be missing some benefits.
```

File: packages/js-client/GUIDE.md [hunk 18]
```diff
@@ -0,0 +1,448 @@
...
+- Contracts
+- JS Client (optional)
+- Subgraph (optional)
+- UI (optional)
+
+The JS client sits as a middle layer between the contracts and the UI. It is responsible for fetching data from the contracts and exposing it to the UI. It also handles the communication between the UI and the contracts.
+
+Using this template and the `@aragon/sdk-client-common` package, you will be able to build a JS client for your plugin in a few minutes.
+
+The `@aragon/sdk-client-common` package provides a set tools to build a JS client for your plugin, it mainly gives you helpers and battle tested logic for handling connections with external services such as:
```
Comment (heueristik) [inline packages/js-client/GUIDE.md]:
```suggestion
The `@aragon/sdk-client-common` package provides a set tools to build a JS client for your plugin. It gives you helpers for handling connections with external services such as:
```

File: packages/js-client/GUIDE.md [hunk 19]
```diff
@@ -0,0 +1,448 @@
...
+
+```bash
+npm install @aragon/sdk-client-common
+// or
+yarn add @aragon/sdk-client-common
+```
+
+Once you have the package installed you can start building your client.
+
+There are 3 main files that you will need to build your client:
```
Comment (heueristik) [inline packages/js-client/GUIDE.md]: Isn't this the wrong approach/writing style? The idea of this template is that everything is provided and already works. 
This guide sounds like you need to do all of that before you can start. But the thing is, that we already created it for the user.

**Review by heueristik — COMMENTED** [review]
I thought this task is related on how to use this template (which files to adapt in this repo to quickly develop your own plugin client).
Basically, adding to and matching the style of [`./USAGE_GUIDE.md`](https://github.com/aragon/osx-plugin-template-hardhat/blob/main/USAGE_GUIDE.md) and populating the missing part in the [`./README.md`](https://github.com/aragon/osx-plugin-template-hardhat/blob/main/README.md).

From what I've read so far (I didn't read the two guides to the end), it now sounds more like a start-from-scratch guide.

This seems too long and repeats the steps that the template already did for the user. Let's clarify if I got something wrong.
Also, I recommend installing language-tool or grammarly to find and correct more spelling and grammar errors than I've found so far.

**Comment by heueristik** [comment]
Closed because of https://github.com/aragon/osx-plugin-template-hardhat/pull/31

**Pull request closed without merging**